subsys/net/buf.c

```diff
@@ -354,22 +430,46 @@ struct net_buf *net_buf_ref(struct net_buf *buf)
 
 struct net_buf *net_buf_clone(struct net_buf *buf, s32_t timeout)
 {
+	u32_t alloc_start = k_uptime_get_32();
 	struct net_buf_pool *pool;
 	struct net_buf *clone;
 
 	NET_BUF_ASSERT(buf);
 
 	pool = net_buf_pool_get(buf->pool_id);
 
-	clone = net_buf_alloc(pool, timeout);
+	clone = net_buf_alloc_len(pool, 0, timeout);
 	if (!clone) {
 		return NULL;
 	}
 
-	net_buf_reserve(clone, net_buf_headroom(buf));
+	/* If the pool supports data referencing use that. Otherwise
+	 * we need to allocate new data and make a copy.
+	 */
+	if (pool->alloc->cb->ref) {
+		clone->__buf = data_ref(buf, buf->__buf);
+		clone->data = buf->data;
+		clone->len = buf->len;
+		clone->size = buf->size;
+	} else {
+		size_t size = buf->size;
 
-	/* TODO: Add reference to the original buffer instead of copying it. */
-	memcpy(net_buf_add(clone, buf->len), buf->data, buf->len);
+		if (timeout != K_NO_WAIT && timeout != K_FOREVER) {
+			u32_t diff = k_uptime_get_32() - alloc_start;
+
+			timeout -= min(timeout, diff);
+		}
+
+		clone->__buf = data_alloc(clone, &size, timeout);
```

Comment:
Isn't it so that if we clone: we expect to clone all the data, so size must be the same as buf->size. If not, cloning failed.